File: website/src/services/questoes.service.ts
import { Filter, Question } from '@/lib/supabase'
import { supabase } from '@/lib/supabase'

interface FiltrosQuestoes {
  disciplina?: string[]
  banca?: string[]
  ano?: string[]
  search?: string
  cargo?: string
  assunto?: string
  instituicao?: string
  regiao?: string
  areaAtuacao?: string
  modalidade?: string
  dificuldade?: string
}

export interface Questao {
  id: string
  professor_id: string
  disciplina: string
  assunto: string
  autor?: string
  isInedita: boolean
  ano: string
  banca: string
  orgao?: string
  textoApoio?: string
  imagemPath?: string
  pergunta: string
  alternativas: string[]
  resposta: string
  explicacao: string
  isPublic: boolean
  createdAt?: Date
  updatedAt?: Date
}

export class QuestoesService {
  static async carregarQuestoes(filtros: Filter): Promise<Question[]> {
    try {
      let query = supabase
        .from('questions')
        .select(`
          *,
          alternatives (*),
          comments (
            *,
            user (
              name,
              avatar_url
            )
          ),
          statistics (*)
        `)
        .order('created_at', { ascending: false })

      if (filtros.discipline) {
        query = query.eq('discipline', filtros.discipline)
      }
      if (filtros.bank) {
        query = query.eq('bank', filtros.bank)
      }
      if (filtros.year) {
        query = query.eq('year', filtros.year)
      }
      if (filtros.search) {
        query = query.ilike('content', `%${filtros.search}%`)
      }

      const { data, error } = await query

      if (error) throw error

      return data as Question[]
    } catch (error) {
      console.error('Erro ao carregar questões:', error)
      return []
    }
  }

  static async carregarFiltros() {
    try {
      const { data: questions, error } = await supabase
        .from('questions')
        .select('discipline, bank, year')

      if (error) throw error

      const disciplines = [...new Set(questions.map(q => q.discipline))].sort()
      const banks = [...new Set(questions.map(q => q.bank))].sort()
      const years = [...new Set(questions.map(q => q.year))].sort((a, b) => b - a)

      return {
        disciplines,
        banks,
        years
      }
    } catch (error) {
      console.error('Erro ao carregar filtros:', error)
      return {
        disciplines: [],
        banks: [],
        years: []
      }
    }
  }

  async getQuestaoPorId(id: string): Promise<Questao | null> {
    try {
      const { data, error } = await supabase
        .from('questions')
        .select()
        .eq('id', id)
        .single();

      if (error) throw error;

      return this.mapQuestao(data);
    } catch (error) {
      console.error('Erro ao buscar questão por ID:', error);
      return null;
    }
  }

  async getQuestaoNumero(id: string): Promise<number> {
    try {
      const { data, error } = await supabase
        .from('questions')
        .select('id')
        .order('created_at', { ascending: true });

      if (error) throw error;

      const index = data.findIndex((q) => q.id === id);
      return index + 1;
    } catch (error) {
      console.error('Erro ao buscar número da questão:', error);
      return 0;
    }
  }

  private mapQuestao(data: any): Questao {
    return {
      id: data.id,
      professor_id: data.professor_id,
      disciplina: data.disciplina,
      assunto: data.assunto,
      autor: data.autor,
      isInedita: data.is_inedita || false,
      ano: data.ano,
      banca: data.banca,
      orgao: data.orgao,
      textoApoio: data.texto_apoio,
      imagemPath: data.imagem_url,
      pergunta: data.pergunta,
      alternativas: data.alternativas,
      resposta: data.resposta,
      explicacao: data.explicacao,
      isPublic: data.is_public || false,
      createdAt: data.created_at ? new Date(data.created_at) : undefined,
      updatedAt: data.updated_at ? new Date(data.updated_at) : undefined,
    };
  }
}

export default new QuestoesService(); 

File: lib/services/questoes_service.dart
import 'dart:convert';
import 'package:flutter/services.dart';
import '../models/questao.dart';
import 'dart:io';
import 'package:supabase_flutter/supabase_flutter.dart';
import 'package:uuid/uuid.dart';

class QuestoesService {
  final _supabase = Supabase.instance.client;

  Future<List<Questao>> carregarQuestoes() async {
    try {
      final user = _supabase.auth.currentUser;
      final response = await _supabase
          .from('questions')
          .select()
          .or('is_public.eq.true,professor_id.eq.${user?.id}')
          .order('created_at', ascending: false);

      return (response as List).map((json) => Questao.fromJson(json)).toList();
    } catch (e) {
      print('Erro ao carregar questões: $e');
      return [];
    }
  }

  Future<Questao?> getQuestaoPorId(String id) async {
    try {
      final response =
          await _supabase.from('questions').select().eq('id', id).single();

      return Questao.fromJson(response);
    } catch (e) {
      print('Erro ao buscar questão por ID: $e');
      return null;
    }
  }

  Future<bool> salvarQuestao(Questao questao) async {
    try {
      String? imageUrl;
      if (questao.imagemPath != null) {
        final file = File(questao.imagemPath!);
        final fileExt = questao.imagemPath!.split('.').last;
        final fileName =
            '${questao.id}_${DateTime.now().millisecondsSinceEpoch}.$fileExt';
        final path = fileName;

        try {
          print('Iniciando upload da imagem...'); 
          await _supabase.storage.from('questoes').upload(path, file);
          print('Upload concluído. Gerando URL pública...'); 

          imageUrl = _supabase.storage.from('questoes').getPublicUrl(path);
          print('URL pública gerada: $imageUrl'); 
        } catch (e) {
          print('Erro no upload da imagem: $e');
        }
      }

      print('Salvando questão no banco com URL da imagem: $imageUrl');

      await _supabase.from('questions').insert({
        'id': questao.id,
        'professor_id': _supabase.auth.currentUser!.id,
        'disciplina': questao.disciplina,
        'assunto': questao.assunto,
        'pergunta': questao.pergunta,
        'alternativas': questao.alternativas,
        'resposta': questao.resposta,
        'explicacao': questao.explicacao,
        'banca': questao.banca,
        'ano': questao.ano,
        'imagem_url': imageUrl,
        'is_inedita': questao.isInedita,
        'autor': questao.autor,
        'texto_apoio': questao.textoApoio,
        'is_public': questao.isPublic,
        'created_at': DateTime.now().toIso8601String(),
        'updated_at': DateTime.now().toIso8601String(),
      });

      return true;
    } catch (e) {
      print('Erro ao salvar questão: $e');
      return false;
    }
  }

  Future<List<Questao>> buscarQuestoes({bool apenasPublicas = false}) async {
    try {
      final user = _supabase.auth.currentUser;
      final query = _supabase.from('questions').select();

      if (apenasPublicas) {
        query.eq('is_public', true);
      } else if (user != null) {
        query.or('is_public.eq.true,professor_id.eq.${user.id}');
      } else {
        query.eq('is_public', true);
      }

      final data = await query;
      return data.map((json) => Questao.fromJson(json)).toList();
    } catch (e) {
      print('Erro ao buscar questões: $e');
      return [];
    }
  }

  String gerarNovoId() {
    return const Uuid().v4();
  }

  Future<int> getQuestaoNumero(String questaoId) async {
    try {
      final response = await _supabase
          .from('questions')
          .select('id')
          .order('created_at', ascending: true);

      final questoes = response as List;
      final index = questoes.indexWhere((q) => q['id'] == questaoId);
      return index + 1;
    } catch (e) {
      print('Erro ao buscar número da questão: $e');
      return 0;
    }
  }

  Future<int> getTotalQuestoes() async {
    try {
      final questoes = await carregarQuestoes();
      return questoes.length;
    } catch (e) {
      print('Erro ao obter total de questões: $e');
      return 0;
    }
  }

  Future<List<String>> getAvailableFilters() async {
    try {
      final questoes = await carregarQuestoes();
      final Map<String, Set<String>> filters = {
        'Disciplina': questoes.map((q) => q.disciplina).toSet(),
        'Assunto': questoes.map((q) => q.assunto).toSet(),
        'Banca': questoes.map((q) => q.banca ?? '').where((b) => b.isNotEmpty).toSet(),
        'Órgão': questoes.map((q) => q.orgao ?? '').where((o) => o.isNotEmpty).toSet(),
        'Ano': questoes.map((q) => q.ano ?? '').where((a) => a.isNotEmpty).toSet(),
        'Autor': questoes.map((q) => q.autor ?? '').where((a) => a.isNotEmpty).toSet(),
      };

      final List<String> availableFilters = [];
      
      filters.forEach((key, values) {
        if (values.isNotEmpty) {
          availableFilters.add(key);
        }
      });
      
      // Add static filters that don't depend on field values
      availableFilters.add('Tipo'); // Para filtrar questões inéditas/públicas
      
      return availableFilters;
    } catch (e) {
      print('Erro ao obter filtros disponíveis: $e');
      return [];
    }
  }
}
